tests/unit/test_run.py

```diff
@@ -0,0 +1,309 @@
+
+
+def test_retrieve_or_create_index_metadata_yaml_create(tmp_path: Path, index_file_content: str):
+    """
+    arrange: given directory with metadata.yaml without docs defined, discourse client that returns
+        url and index file that has been created
+    act: when retrieve_or_create_index is called with that directory and with create_if_not_exists
+        True
+    assert: then create topic is called with the titleised charm name and with placeholder content
+        and the url returned by the client and placeholder content is returned.
+    """
+    metadata_yaml = tmp_path / "metadata.yaml"
+    metadata_yaml.write_text("name: charm-name", encoding="utf-8")
+    mocked_server_client = mock.MagicMock(spec=discourse.Discourse)
+    url = "http://server/index-page"
+    mocked_server_client.create_topic.return_value = url
+
+    returned_page = run.retrieve_or_create_index(
+        create_if_not_exists=True, base_path=tmp_path, server_client=mocked_server_client
+    )
+
+    assert returned_page.url == url
+    assert index_file_content in returned_page.content.lower()
+
+    mocked_server_client.create_topic.assert_called_once()
+    call_kwargs = mocked_server_client.create_topic.call_args.kwargs
+    assert "title" in call_kwargs and "Charm Name" in call_kwargs["title"]
+    assert "content" in call_kwargs and index_file_content in call_kwargs["content"]
+
+
+def test_retrieve_or_create_index_metadata_yaml_retrieve_discourse_error(tmp_path: Path):
+    """
+    arrange: given directory with metadata.yaml with docs defined and discourse client that
+        raises DiscourseError
+    act: when retrieve_or_create_index is called with that directory
+    assert: then ServerError is raised.
+    """
+    metadata_yaml = tmp_path / "metadata.yaml"
+    metadata_yaml.write_text("docs: http://server/index-page", encoding="utf-8")
+    mocked_server_client = mock.MagicMock(spec=discourse.Discourse)
+    mocked_server_client.retrieve_topic.side_effect = DiscourseError
+
+    with pytest.raises(ServerError) as exc_info:
+        run.retrieve_or_create_index(
+            create_if_not_exists=False,
+            base_path=tmp_path,
+            server_client=mocked_server_client,
+        )
+
+    assert_substrings_in_string(("index page", "retrieval", "failed"), str(exc_info.value).lower())
+
+
+def test_retrieve_or_create_index_metadata_yaml_retrieve(tmp_path: Path):
+    """
+    arrange: given directory with metadata.yaml with docs defined and discourse client that
+        returns the index page content
+    act: when retrieve_or_create_index is called with that directory
+    assert: then retrieve topic is called with the docs key value and the content returned by the
+        client and docs key is returned.
+    """
+    url = "http://server/index-page"
+    content = "content 1"
+    metadata_yaml = tmp_path / "metadata.yaml"
+    metadata_yaml.write_text(f"docs: {url}", encoding="utf-8")
```

Comment:
This is repeated multiple times, could be interesting to factor this in a method ?

Reply:
Thanks, this should be fixed in the next commit